infrastructure/comms/src/connection/zmq/context.rs

```diff
@@ -0,0 +1,54 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use zmq;
+
+use crate::connection::{types::SocketType, zmq::ZmqError};
+
+#[derive(Clone)]
+pub struct Context(zmq::Context);
+
+impl Context {
```

Comment:
Is this (very thin) abstraction to avoid ZMQ being explicitly present in the Domain layer?

Reply:
There will be  a number of components which will use one context, probably not set up on the domain layer, but still I'd like to keep zmq to a small part of the code and abstracted from other components which set up peer connections - Context could touch a lot of public struct/functions, and I'm thinking it's better to have a `tari_comms` context passed around than a `rust-zmq` Context.  Still not 100% decided on this though. Think we'll give up on the idea quickly if we start looking like we're having to implement more abstractions on an abstraction of an abstraction just to make this fit.

Reply:
Cool.